### PR TITLE
Add a wrapText option to table cell renderer

### PR DESCRIPTION
**Pull request by gscshoyru**
#### Checklist

- [x] [Enable CircleCI for your fork](https://circleci.com/add-projects)
- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

Adds a `wrapText` prop to cells -- off by default (current behavior). If enabled, will wrap text onto further lines. This text can be seen by resizing the row (or will overflow the cell if truncation is disabled). 

Also, made it so that the `TruncatedFormatter` will still show the popover thing if the cell is being overflowed in either direction (out the bottom or to the right). 

#### Reviewers should focus on:

If my logic/css logic makes sense. 

If I should be calling it "wrapText" as I currently am, or "nowrap" as many other things seem to do. 

Note: I haven't written tests yet, just fixed the ones that were breaking. 

#### Screenshot

mar-08-2017 17-03-38

### CODE REVIEW

**Comment by helenkg** [comment]
can I automatically wrap text and make each cell (ie row) as tall as it needs to be to show all text? and adjust height as I drag width of column?

**Comment by tgreenwatts** [comment]
Sorry no, not right now, but at least you can make the rows big enough & wrap the text (2 steps) and see them

**Comment by helenkg** [comment]
So, it will use as many rows as the row is tall? 
And it will use more or less rows I drag row height?

**Comment by gscshoyru** [comment]
All this is doing is adding a flag on the cell that will make it wrap text rather than being a single line. (Before this, you could resize row heights but it didn't do anything useful -- it still displayed content on a single line.) You still need to resize the row to view the text inside, however -- this commit doesn't add any logic around resizing row heights or anything like that. The rows will be as tall or as short as you drag them to be, nothing more, nothing less.

**Comment by helenkg** [comment]
we can start here :) still need the auto-resize though

**Comment by llorca** [comment]
The auto-resizing feature will come later on 👍  It's non trivial at the moment and requires a lot of work

**Comment by giladgray** [comment]
@helenkg I wouldn't get your hopes up on the auto-sizing feature. We've discussed at length about how it's not impossible, but it's so computationally expensive as to make it impractical to implement.

File: packages/table/test/formatsTests.tsx [hunk 1]
```diff
@@ -115,7 +119,11 @@ describe("Formats", () => {
                 The fair Ophelia! -- Nymph, in thy orisons
                 Be all my sins remembered.
             `;
-            const comp = harness.mount(<TruncatedFormat truncateLength={0}>{str}</TruncatedFormat>);
+            const comp = harness.mount(
+                <div className={Classes.TABLE_NO_WRAP_TEXT}>
+                    <TruncatedFormat truncateLength={0}>{str}</TruncatedFormat>
+                </div>,
```
Comment (giladgray) [inline packages/table/test/formatsTests.tsx]: this is an awkward change. does `TruncatedFormat` by itself no longer work as expected? cuz that seems problematic (although it's unlikely to be used directly on its own like this).

Comment (gscshoyru) [inline packages/table/test/formatsTests.tsx]: Hmmm... 

So what does "TruncatedFormat" on its own mean? My reading/understanding of its behavior is that if detectTruncation is true (on by default) then essentially all it is doing is detecting whether or not it is truncated (as constrained by its parent) and if so is displaying the three-dot popover button. 

I'm going to add more tests that test the same behavior with vertical truncation as well as horizontal -- but I'm not sure it makes sense for TruncatedFormat to truncate on its own if detectTrunctation is turned on. If this is incorrect what would you rather the behavior be, here?

Comment (giladgray) [inline packages/table/test/formatsTests.tsx]: I'll leave the final decision up to you as you know the codebase better.

Comment (themadcreator) [inline packages/table/test/formatsTests.tsx]: I'm actually totally okay with this. The idea is that truncatedFormat can live in both a wrapped and unwrapped div (like a `<Cell>`. So, this test actually confirms that.

**Comment by adidahiya** [comment]
@gscshoyru @llorca what should the default end-user experience for this be like? a column header context menu that toggles `wrapText` for the column's cells?

**Comment by gscshoyru** [comment]
Depends on what you mean by default. The table doesn't currently ship with anything like that (I'm not sure it could at present), but it certainly would be a good way to do it, and one I would suggest. Because the wrapping is controlled by a prop on the cell, which is in turn controlled by the cell renderer, consumers of the library could turn on wrapping on cells or rows or columns in any way they desire, and control it however they'd like. But as set up at present, we don't track anything like that in the table itself -- it needs to be tracked and controlled externally. If you think we should be doing something with this inside the table itself, we may need to rethink this a bit.

**Comment by adidahiya** [comment]
I know that I'm not getting this for free from the table out of the box. I'm asking how I should expose this in an application if I want this available to end users _today_ given the primitives that we have with this PR and #802. Mostly a question for @llorca (same goes for what to bake into `<Table>` out of the box).

**Comment by llorca** [comment]
@adidahiya +1 on enabling/disabling wrapping from column header action dropdown, also would like to have this generally available in right click context menus

**Comment by llorca** [comment]
Also, even though it may not be ideal for the use case we're targeting, are we in agreement with `wrapText` defaulting to `false`?

**Comment by gscshoyru** [comment]
If we don't default it to false it'd be a breaking change -- and and excel and google spreadsheets do single line by default as well -- it it's probably the right default choice.

**Comment by helenkg** [comment]
:( google docs and excel know how to set the height automatically if you configure a column to wrap text

**Comment by tgreenwatts** [comment]
Yep issue isn't so much that we don't know what the heights are, but that we'd have to draw everything to figure that out - and the point of our table code was that we are virtually rendering and don't have to do that. There are improvements we can make to make it seem like we're wrapping & expanding automatically, but we're not there yet

**Comment by adidahiya** [comment]
@helenkg Google Docs and Excel were also built with more than ~2 developers :joy:

**Comment by brsanthu** [comment]
Thank you @gscshoyru for this PR. Awesome feature.

**Pull request merged**